synapse/replication/http/send_event.py

```diff
@@ -92,8 +114,23 @@ def __init__(self, hs):
         self.store = hs.get_datastore()
         self.clock = hs.get_clock()
 
+        # The responses are tiny, so we may as well cache them for a while
+        self.response_cache = ResponseCache(hs, timeout_ms=30 * 60 * 1000)
+
+    def on_PUT(self, request, event_id):
+        result = self.response_cache.get(event_id)
+        if not result:
+            result = self.response_cache.set(
+                event_id,
+                self._handle_request(request)
+            )
+        else:
+            logger.warn("Returning cached response")
```

Comment:
I wouldn't mind if you logged the state of `result.called` (ie whether we're returning a completed response or are still waiting for the request to complete), but this is fine.